src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs

```diff
@@ -230,8 +232,62 @@ internal bool TryGetPropertyStoreItem(string propertyName, [NotNullWhen(true)] o
                     propertyStoreItems[property.Name] = item;
                 }
 
+                return CombineMetadataTypeAttributes(propertyStoreItems);
+            }
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:DynamicallyAccessedMembersAttribute",
+                       Justification = "The ctor is marked with DynamicallyAccessedMembers.")]
+            private Dictionary<string, PropertyStoreItem> CombineMetadataTypeAttributes(Dictionary<string, PropertyStoreItem> propertyStoreItems)
+            {
+                // check if metadata attribute is present on the type
+                var metadataAttr = _type.CustomAttributes.FirstOrDefault(attr => attr.AttributeType.Name == nameof(MetadataTypeAttribute));
+                if (metadataAttr != null)
+                {
+                    // get properties of metadata type
+                    TypeInfo t = (TypeInfo)metadataAttr.ConstructorArguments.First().Value!;
+                    var properties = t.GetProperties()
+                        .Where(prop => !prop.GetIndexParameters().Any());
+
+                    foreach (PropertyInfo property in properties)
+                    {
+                        // metadata property name must match an existing property
+                        if (!propertyStoreItems.TryGetValue(property.Name, out PropertyStoreItem? baseProperty))
+                        {
+                            // re-used existing description since it is the same exception
+                            throw new InvalidOperationException(SR.Format(SR.AssociatedMetadataTypeTypeDescriptor_MetadataTypeContainsUnknownProperties,
+                                _type.FullName,
+                                property.Name));
+                        }
+
+                        var mergedAttributes = MergeAttributes(baseProperty!.ValidationAttributes.Cast<Attribute>(),
```

Comment:
I don't believe you need a `!` here on `baseProperty!`. The dictionary doesn't contain `null`s, and `TryGetValue` returned `true`.

Reply:
VS used to complain it was potentially null. It doesn't now. Thanks